### ISSUE TITLE
Allow killing the TestKit daemon

### ISSUE THREAD

**Issue by dansanduleac** [gradle/gradle#12535]
### Use case

I want to run gradle integration tests in parallel using JUnit 5.

This is possible to instrument via TestKit, as long as we pass each test instance its own directory via `GradleRunner.withTestKitDir`. 

However, as pointed out in https://github.com/gradle/gradle/issues/3093#issuecomment-336523391 as well, the GradleRunner API will spin up a daemon with a 3 minute timeout when it creates the connector: https://github.com/gradle/gradle/blob/a9f43feec842fb15dde13b9058af09be98d78f12/subprojects/test-kit/src/main/java/org/gradle/testkit/runner/internal/ToolingApiGradleExecutor.java#L192

This means that if I have junit-jupiter create temporary dirs for each testkit invocation, it will try to delete them at the end of the test, and sometimes fail, because the daemon is still alive and sometimes messes with files while the directory is being deleted.

### Expected Behavior

I'd like to be able to programmatically kill the testkit daemon for a given testkit directory, via some API on the `GradleRunner`.

### Current Behavior

I can't kill the testkit daemon, and so can't reliably run gradle integration tests in parallel, unless I accept that they will leak temporary files and directories.

**Comment by szpak** [comment]
I've bumped into that issue accidentally looking for something else, however, maybe the new API from [GradleConnector](https://docs.gradle.org/6.5-rc-1/release-notes.html#improvements-for-tooling-providers) could be used as a workaround (just an idea, I haven't checked if `GradleRunner` allows to use/get `GradleConnector` easily)?

**Comment by dansanduleac** [comment]
Sadly it hides the GradleConnector from you, but with that change they recently made, it should be much easier to add a function to the `GradleRunner`!

**Comment by Tapchicoma** [comment]
I see that test daemons are stopped after all tests are completed via this [hook](https://github.com/gradle/gradle/blob/34c45c143c43eb3f4ac98468a9b1d17575bcae30/subprojects/test-kit/src/main/java/org/gradle/testkit/runner/internal/ToolingApiGradleExecutor.java#L97). Probably as a workaround you could add a new Gradle clean task that will clean TestKit dir after test task via `finalizedBy`.

But this solution is not working for the cases when test need to run sub-sequent build on a fresh daemon. So, imho, exposing daemon idle timeout parameter would be nice to have - then test could set this timeout to `0` and wait for a short time for daemon to stop before starting next build.

**Comment by cruzel89** [comment]
I ran into this issue as well the last days using Gradle v7.4 (on Windows 10 x64 with openjdk v11.0.12).

In my case jUnit tests for an own developed Gradle plugin are failing due to this, because the [@TempDir](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-built-in-extensions-TempDirectory) annotated temporary folder which is used as temporary `gradle.user.home` for the `GradleRunner` of the [TestKit library](https://docs.gradle.org/current/userguide/test_kit.html) cannot be automatically deleted after the jUnit test was executed:

```
java.io.IOException: Failed to delete temp directory C:\Users\****\AppData\Local\Temp\junit492549456655448575. The following paths could not be deleted (see suppressed exceptions for details):
gradlehome,
gradlehome\caches,
gradlehome\caches\7.4,
gradlehome\caches\7.4\executionHistory,
gradlehome\caches\7.4\executionHistory\executionHistory.lock,
gradlehome\caches\7.4\fileContent,
gradlehome\caches\7.4\fileContent\fileContent.lock,
gradlehome\caches\7.4\fileHashes,
gradlehome\caches\7.4\fileHashes\fileHashes.bin,
gradlehome\caches\7.4\fileHashes\fileHashes.lock,
gradlehome\caches\7.4\fileHashes\resourceHashesCache.bin,
gradlehome\caches\jars-9,
... (some project-internal jars removed)
gradlehome\caches\jars-9\jars-9.lock,
gradlehome\caches\journal-1,
gradlehome\caches\journal-1\file-access.bin,
gradlehome\caches\journal-1\journal-1.lock,
gradlehome\caches\modules-2,
gradlehome\caches\modules-2\metadata-2.97,
gradlehome\caches\modules-2\metadata-2.97\module-artifact.bin,
gradlehome\caches\modules-2\metadata-2.97\module-artifacts.bin,
gradlehome\caches\modules-2\metadata-2.97\module-metadata.bin,
gradlehome\caches\modules-2\metadata-2.97\resource-at-url.bin,
gradlehome\caches\modules-2\modules-2.lock
```

Unfortunately the `GradleRunner` of TestKit neither seems to run without Daemon when using `.withDebug(true)` (as proposed [here](https://discuss.gradle.org/t/testkit-how-to-turn-off-daemon/17843/2)) nor when applying `.withArguments("-Dorg.gradle.daemon=false")`. In both cases jUnit still fails to automatically delete the temp directory with an `java.io.IOException` after the test cases have been executed. Is there a different way of using `GradleRunner` without Daemon?

The workaround in Groovy I am using now is to create the temporary folder manually before the test execution via
```groovy
testProjectDir = Files.createTempDirectory("junit")
```
and force-delete it after the jUnit tests with:
```groovy
FileUtils.forceDeleteOnExit(testProjectDir.toFile())
```

But I would really appreciate a working solution using the [@TempDir](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-built-in-extensions-TempDirectory) annotated folder functionality of jUnit 5.

**Comment by ghillert** [comment]
Just going through the pain of getting tests to work on Windows. Interesting to see that this issue is still open 4 years later. Adding a separate "clean" task in my integration tests using `GradleRunner` seems to get most tests pass, but I have an annoying test that involves a multi-module test project where nothing seems to work including the above work-around of using `FileUtils.forceDeleteOnExit`. I think, users should really be able to execute a build **once** using GradleRunner and have all handles released.

**Comment by ljacomet** [comment]
Sorry for the late reply

This feature request is in the backlog of the relevant team and is prioritized by them.

***

We understand that TestKit needs improvement, but this is not actively worked on at this time.

As indicated in some answers above, there are possible workarounds for some scenarios.
